beacon/blockchain/process_proposal.go

```diff
@@ -90,24 +93,41 @@ func (s *Service) ProcessProposal(
 	}
 
 	blk := signedBlk.GetBeaconBlock()
+
+	// There are two different timestamps:
+	//     - The "consensus time" is determined by CometBFT consensus and can be retrieved with `req.GetTime()`
+	//     - The "block time" is determined by beacon-kit consensus and can be retrieved with `blk.GetTimestamp()`
+	// The "consensus time" is what the network agrees the current time is based on CometBFT PBTS.
+	// This "consensus time" is used to constrain the timestamp set as the "block time" by the
+	// beacon-kit app, but they are not always equal in value. The "block time" is used by the
+	// beacon-kit consensus and execution layers to determine the active fork version.
+	//
+	// When unmarshaling the BeaconBlock, we do not yet have access to the "block time", so we
+	// must rely on the "consensus time" as our best estimation of the "block time" needed to
+	// determine the current fork version. Since the two timestamps could be different, we need to
+	// ensure that the fork version for these timestamps are the same. This may result in a failed
+	// proposal or two at the start of the fork.
+	blkVersion := s.chainSpec.ActiveForkVersionForTimestamp(blk.GetTimestamp())
```

Comment:
general question: I wonder if instead of guessing, we should explicitly add a third transaction to the cometBFT block to communicate metadata like the fork version (for the time being) and possibly others in the next future.
It should simplify the flow I believe as:
- we would check the fork against the timestamps (which is a cheap check).
- we would correctly initialize the fork version in the blocks (which is never serialized afaik, nor inferred by its content)

Reply:
My other question whether we could infer the version from the data serialized/not serialized somehow, which also would avoid the possible liveness issue.